File: Lee/Baekjoon Online Judge/20154번: 이 구역의 승자는 누구야?!/20154_2.cpp
#include <cstddef>
#include <cstdint>
#include <iterator>
#include <algorithm>
#include <numeric>
#include <array>
#include <vector>
#include <string>
#include <iostream>

inline auto map(char c)
{
	static constexpr std::array<std::uint_fast32_t, 26> numbers = {
		3, 2, 1, 2, 3, 3, 3, 3, 1, 1, 3, 1, 3,
		3, 1, 2, 2, 2, 1, 2, 1, 1, 2, 2, 2, 1
	};

	return numbers[static_cast<std::size_t>(c - 'A')];
}

int main()
{
	std::string line;
	std::getline(std::cin, line);

	std::vector<unsigned short int> scoreboard;
	scoreboard.reserve(line.length());

	std::transform(
		std::cbegin(line),
		std::cend(line),
		std::back_inserter(scoreboard),
		map
	);

	const auto sum = std::reduce(
		std::cbegin(scoreboard),
		std::cend(scoreboard),
		std::uint_fast32_t{}
	);

	if (sum % 2 == 1)
		std::cout << "I'm a winner!\n";
	else
		std::cout << "You're the winner?\n";

	return 0;
}

File: Lee/Baekjoon Online Judge/20154번: 이 구역의 승자는 누구야?!/20154_2.rs
use std::io::prelude::*;
use std::io::stdin;

#[inline]
fn map(c: char) -> u32 {
    match c {
        'A' | 'E'..='H' | 'K' | 'M'..='N' => 3,
        'B' | 'D' | 'P'..='R' | 'T' | 'W'..='Y' => 2,
        'C' | 'I'..='J' | 'L' | 'O' | 'S' | 'U'..='V' | 'Z' => 1,
        _ => 0,
    }
}

fn main() {
    let sum: u32 = {
        let mut input = stdin().lock();
        let mut line = String::new();

        input.read_line(&mut line).unwrap();
        line.trim_end().chars().map(map).sum()
    };

    match sum % 2 {
        1 => {
            println!("I'm a winner!");
        }
        _ => {
            println!("You're the winner?");
        }
    }
}
